src/PortAudio.jl

```diff
@@ -91,83 +93,50 @@ struct PortAudioStream{T}
     recover_xruns::Bool
     sink_buffer::Buffer{T}
     source_buffer::Buffer{T}
+end
 
-    # this inner constructor is generally called via the top-level outer
-    # constructor below
-
-    # TODO: pre-fill outbut buffer on init
-    # TODO: recover from xruns - currently with low latencies (e.g. 0.01) it
-    # will run fine for a while and then fail with the first xrun.
-    # TODO: figure out whether we can get deterministic latency...
-    function PortAudioStream{T}(
-        indev::PortAudioDevice,
-        outdev::PortAudioDevice,
-        inchans,
-        outchans,
-        sr,
-        latency,
-        warn_xruns,
-        recover_xruns,
-    ) where {T}
-        inchans = inchans == -1 ? indev.maxinchans : inchans
-        outchans = outchans == -1 ? outdev.maxoutchans : outchans
-        inparams = if (inchans == 0)
-            Ptr{Pa_StreamParameters}(0)
-        else
-            Ref(Pa_StreamParameters(indev.idx, inchans, type_to_fmt[T], latency, C_NULL))
-        end
-        outparams = if (outchans == 0)
-            Ptr{Pa_StreamParameters}(0)
-        else
-            Ref(Pa_StreamParameters(outdev.idx, outchans, type_to_fmt[T], latency, C_NULL))
-        end
-        # finalizer(close, this)
-        pointer_ref = @stderr_as_debug Pa_OpenStream(
-            inparams,
-            outparams,
-            sr,
-            0,
-            paNoFlag,
-            nothing,
-            nothing,
-        )
-        sink_buffer = Buffer{T}(outdev, outchans)
-        source_buffer = Buffer{T}(indev, inchans)
-        Pa_StartStream(pointer_ref[])
-        this = new(
-            sr, 
-            latency, 
-            pointer_ref, 
-            warn_xruns,
-            recover_xruns,
-            sink_buffer,
-            source_buffer
+function make_parameters(device, channels, T, latency, host_api_specific_stream_info)
+    if channels == 0
+        Ptr{Pa_StreamParameters}(0)
+    else
+        Ref(
+            Pa_StreamParameters(
+                device.idx,
+                channels,
+                type_to_fmt[T],
+                latency,
+                convert_nothing(host_api_specific_stream_info),
+            ),
         )
-        # pre-fill the output stream so we're less likely to underrun
-        prefill_output(this.sink)
+    end
+end
 
-        this
+function fill_max_channels(channels, bounds)
+    if channels === max
```

Comment:
avoid code duplication. also node that i've replaced the undocuented -1 sentinel and instead you can pass the max function itself. this seems clearer to me, if we like the change i can document it